### PR TITLE
From Black to Ruff

### PR DESCRIPTION
**Pull request by Avasam**
This PR explores using Ruff format in place of Black.

The main advantages are:
- speed (there is a small, but noticeable difference when running pre-commit on the entire repo, for example)
- Simplified configs and docs (only because we already use Ruff)
- 1 less dependency
- Arguably better formatting in some areas (especially where a comment previously prevented a line-wrap)

Eventually scripts could also be simplified as Ruff would be run using a single command https://github.com/astral-sh/ruff/issues/8232

Disadvantages:
- Likely not yet as "stable" as Black across versions
- Long standing collaboration between typeshed and Black (and overlap in maintainers)
- wrapping of lines with `type` comments may be undesirable to some https://github.com/astral-sh/ruff/issues/10516

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,6 +14,15 @@ exclude = [
...
 ]
 
+[tool.ruff.format]
+skip-magic-trailing-comma = true
+exclude = [
+    # Exclude protobuf files because they have long line lengths
+    # Ideally, we could configure Ruff to allow longer line lengths
+    # for just these files, by adding a ruff.toml config file at their root.
+    "**_pb2.pyi",
+]
```
Comment (Avasam) [inline pyproject.toml:24]: Ruff is so fast and fixers are already run on these files as part of the generator scripts, I feel like this isn't even necessary.
I just kept if for now to have an apples-to-apples comparison.

**Comment by zanieb** [comment]
Exciting to see this! Note we added the formatter to our [versioning policy](https://docs.astral.sh/ruff/versioning/) and it should be quite stable!

**Comment by JelleZijlstra** [comment]
I don't see much advantage in this change, though as a maintainer of Black I might be biased. I don't think formatting speed is a bottleneck for us.

**Comment by hauntsaninja** [comment]
Having to manually move type comments around isn't a good sign. Black basically just totally leaves those alone. I don't think there's much advantage here. Also note that ruff formatter currently relies on Black for its style. Black has historically collaborated with typeshed to determine Black's pyi style, so there's sort of a healthy collaboration / feedback loop there.

**Comment by Avasam** [comment]
Thanks everyone for your feedback and additional context. I had an inkling about a sentiment towards keeping black, but I couldn't be sure without trying or asking.

I'll close this PR, but if you have any other opinion you'd like to share, feel free to do so here.

**Pull request closed without merging**